### PR TITLE
Removes data fetch related logic

### PR DESCRIPTION
**Pull request by velioglu**
DESCRIPTION: Removes shard data fetch related logic 

This PR removes the codes related to the data fetch. 

On the code side, following items are removed.

- Remove adding `SHARD_FETCH_TASK` and related functions from physical planner.
- Remove fetching regular tables within task-tracker and real-time executor
- Remove implementation of `worker_fetch_regular_table`, `worker_fetch_foreign_file` and functions used only by these two.
- Remove `master_expire_table_cache` and functions used only by that function

Since we have many tests in our regression tests which use broadcast or local join logic, test files are updated to make them runnable without these logics. While changing tests following approaches followed :

- If there is a small append table convert it to the reference table.
- If above conversion causes a problem, replicate shard of append tables to each worker node.
- If we have a join with two large tables and that join utilized broadcast or local join logic, convert it to the re-partition join.
- If can't use re-partition join, convert query to router plannable query.

### CODE REVIEW

File: src/backend/distributed/citus--7.2-3--7.2-4.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/* citus--7.2-2--7.2-3 */
+
+DROP FUNCTION IF EXISTS master_expire_table_cache(regclass);
```
Comment (marcocitus) [inline src/backend/distributed/citus--7.2-3--7.2-4.sql]: Be careful with these. We messed up before by using the wrong signature and due to the IF EXISTS this is silently ignored.
E.g. Does this actually drop the function without the pg_catalog prefix?

You may also want to remove the original create function from the SQL file, such that we could safely remove the C definition.

File: src/backend/distributed/utils/citus_copyfuncs.c [hunk 2]
```diff
@@ -241,7 +241,6 @@ CopyNodeTask(COPYFUNC_ARGS)
 	COPY_SCALAR_FIELD(upstreamTaskId);
 	COPY_NODE_FIELD(shardInterval);
 	COPY_SCALAR_FIELD(assignmentConstrained);
-	COPY_SCALAR_FIELD(shardId);
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/utils/citus_copyfuncs.c]: Actually, this currently crashes without that when asserts enabled:
```
Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x0000000000006900
0x000000010896cd44 in outNode (str=0x7fff575e99b0, obj=0x6900) at outfuncs.c:3601
3601		else if (IsA(obj, List) ||IsA(obj, IntList) || IsA(obj, OidList))
(gdb) bt
#0  0x000000010896cd44 in outNode (str=0x7fff575e99b0, obj=0x6900) at outfuncs.c:3601
#1  0x0000000109422ce4 in OutTask (str=0x7fff575e99b0, raw_node=0x7fb2e202d0e8) at utils/citus_outfuncs.c:453
#2  0x0000000108977d15 in _outExtensibleNode (str=0x7fff575e99b0, node=0x7fb2e202d0e8) at outfuncs.c:2577
#3  0x000000010896da8d in outNode (str=0x7fff575e99b0, obj=0x7fb2e202d0e8) at outfuncs.c:4070
#4  0x000000010896e62e in _outList (str=0x7fff575e99b0, node=0x109398600) at outfuncs.c:182
#5  0x000000010896cd80 in outNode (str=0x7fff575e99b0, obj=0x109398600) at outfuncs.c:3602
#6  0x00000001094224e9 in OutJobFields (str=0x7fff575e99b0, node=0x7fb2e20273d8) at utils/citus_outfuncs.c:317
#7  0x0000000109422465 in OutJob (str=0x7fff575e99b0, raw_node=0x7fb2e20273d8) at utils/citus_outfuncs.c:331
#8  0x0000000108977d15 in _outExtensibleNode (str=0x7fff575e99b0, node=0x7fb2e20273d8) at outfuncs.c:2577
#9  0x000000010896da8d in outNode (str=0x7fff575e99b0, obj=0x7fb2e20273d8) at outfuncs.c:4070
#10 0x0000000109421e4f in OutDistributedPlan (str=0x7fff575e99b0, raw_node=0x10939a5e0) at utils/citus_outfuncs.c:182
#11 0x0000000108977d15 in _outExtensibleNode (str=0x7fff575e99b0, node=0x10939a5e0) at outfuncs.c:2577
#12 0x000000010896da8d in outNode (str=0x7fff575e99b0, obj=0x10939a5e0) at outfuncs.c:4070
#13 0x000000010897c405 in nodeToString (obj=0x10939a5e0) at outfuncs.c:4260
#14 0x00000001093ebab5 in CheckNodeCopyAndSerialization (node=0x10939a5e0) at planner/distributed_planner.c:1079

```

with 
```SQL

SET citus.shard_count TO 4;

DROP TABLE test_int, test_bigint;
CREATE TABLE test_int (a int, b int);
CREATE TABLE test_bigint (a bigint, b bigint);


SELECT create_distributed_table('test_int', 'a');
SELECT create_distributed_table('test_bigint', 'a');

SELECT count(*) FROM test_int, test_bigint WHERE test_int.a = test_bigint.a;
```

File: src/backend/distributed/executor/multi_task_tracker_executor.c [hunk 3]
```diff
@@ -1923,14 +1922,6 @@ ConstrainedNonMergeTaskList(List *taskAndExecutionList, Task *task)
...
 		dependedTaskList = upstreamTask->dependedTaskList;
 	}
-	else if (taskType == SHARD_FETCH_TASK)
-	{
-		List *upstreamTaskList = UpstreamDependencyList(taskAndExecutionList, task);
-		Assert(list_length(upstreamTaskList) == 1);
-
-		upstreamTask = (Task *) linitial(upstreamTaskList);
-		dependedTaskList = upstreamTask->dependedTaskList;
-	}
```
Comment (onderkalaci) [inline src/backend/distributed/executor/multi_task_tracker_executor.c]: Are we still sure that the following assert would never break?

File: src/test/regress/sql/multi_simple_queries.sql [hunk 4]
```diff
@@ -132,10 +132,17 @@ SELECT * FROM articles, position('om' in 'Thomas') ORDER BY 2 DESC, 1 DESC, 3 DE
 SELECT * FROM articles WHERE author_id IN (SELECT id FROM authors WHERE name LIKE '%a');
 
 -- subqueries are supported in FROM clause
+
+SET citus.large_table_shard_count TO 1;
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/test/regress/sql/multi_simple_queries.sql]: To pass broadcast join logic

File: src/test/regress/expected/multi_mx_repartition_udt_prepare.out [hunk 5]
```diff
@@ -156,10 +156,12 @@ INSERT INTO repartition_udt_other values (10, '(2,1)'::test_udt, 'foo');
 INSERT INTO repartition_udt_other values (11, '(2,2)'::test_udt, 'foo');
 INSERT INTO repartition_udt_other values (12, '(2,3)'::test_udt, 'foo');
 SET client_min_messages = LOG;
--- Query that should result in a repartition join on int column, and be empty.
+-- This query was intended to test "Query that should result in a repartition 
+-- join on int column, and be empty." In order to remove broadcast logic, we
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_mx_repartition_udt_prepare.out]: I think `In order to remove broadcast logic` -> `while removing bcast logic` ?

File: src/backend/distributed/executor/multi_task_tracker_executor.c [hunk 6]
```diff
@@ -2104,10 +2080,7 @@ ReassignTaskList(List *taskList)
 		bool transmitCompleted = TransmitExecutionCompleted(taskExecution);
 		if ((task->taskType == SQL_TASK) && transmitCompleted)
 		{
-			List *shardFetchTaskList = ShardFetchTaskList(task->dependedTaskList);
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/backend/distributed/executor/multi_task_tracker_executor.c]: Has been updated

File: src/backend/distributed/executor/multi_task_tracker_executor.c [hunk 7]
```diff
@@ -2104,10 +2080,7 @@ ReassignTaskList(List *taskList)
...
-
 			completedTaskList = lappend(completedTaskList, task);
-			completedTaskList = TaskListUnion(completedTaskList, shardFetchTaskList);
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/backend/distributed/executor/multi_task_tracker_executor.c]: Has been removed

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 8]
```diff
@@ -3765,20 +3763,7 @@ DataFetchTaskList(uint64 jobId, uint32 taskIdIndex, List *fragmentList)
 	foreach(fragmentCell, fragmentList)
 	{
 		RangeTableFragment *fragment = (RangeTableFragment *) lfirst(fragmentCell);
-		if (fragment->fragmentType == CITUS_RTE_RELATION)
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/backend/distributed/planner/multi_physical_planner.c]: Has been updated

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 9]
```diff
@@ -4168,17 +4023,11 @@ PruneSqlTaskDependencies(List *sqlTaskList)
 			Task *dataFetchTask = (Task *) lfirst(dependedTaskCell);
 
 			/*
-			 * If we have a shard fetch task for the anchor shard, or if we have
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/backend/distributed/planner/multi_physical_planner.c]: Has been updated

File: src/include/distributed/pg_dist_shard.h [hunk 10]
```diff
@@ -52,10 +52,9 @@ typedef FormData_pg_dist_shard *Form_pg_dist_shard;
 #define Anum_pg_dist_shard_shardmaxvalue 6
 
 /*
- * Valid values for shard storage types include relay file, foreign table,
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/include/distributed/pg_dist_shard.h]: Has been separated

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/planner/multi_physical_planner.c]: I think the function comment still doesn't make much sense. There is no `shardList` provided to the function. The only fragmentType is `CITUS_RTE_REMOTE_QUERY ` which we could probably reword as (please read and fix any logical/grammatical errors below if exists):
```

/*
 * DataFetchTaskList builds a **merge** fetch task for 
every *remote query result* in the given fragment list, 
appends these **merge fetch** tasks into a list, and returns this list.
 */
```

File: src/backend/distributed/shared_library_init.c [hunk 11]
```diff
@@ -358,19 +358,6 @@ RegisterCitusConfigVariables(void)
 		0,
 		NULL, NULL, NULL);
 
-	DefineCustomBoolVariable(
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c]: well, we shouldn't remove the GUC, no? Should we make `ExpireCachedShards ` a static variable and add a comment here that the GUC is not used anymore?

**Comment by onderkalaci** [comment]
> I think we're ready to merge this PR 

@velioglu  Oh, wait. We should add this version to `multi_extension.sql` test. I think that is a must have and we shouldn't forget to add that.

File: src/test/regress/expected/multi_mx_repartition_udt_prepare.out [hunk 12]
```diff
@@ -156,10 +156,12 @@ INSERT INTO repartition_udt_other values (10, '(2,1)'::test_udt, 'foo');
 INSERT INTO repartition_udt_other values (11, '(2,2)'::test_udt, 'foo');
 INSERT INTO repartition_udt_other values (12, '(2,3)'::test_udt, 'foo');
 SET client_min_messages = LOG;
--- Query that should result in a repartition join on int column, and be empty.
+-- This query was intended to test "Query that should result in a repartition
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/test/regress/expected/multi_mx_repartition_udt_prepare.out]: Will be handled with the following PR.

File: src/test/regress/expected/multi_mx_repartition_udt_w1.out [hunk 13]
```diff
@@ -7,11 +7,12 @@ SET client_min_messages = LOG;
 SET citus.large_table_shard_count = 1;
 SET citus.task_executor_type = 'task-tracker';
 SET citus.log_multi_join_order = true;
--- Query that should result in a repartition join on int column, and be empty.
+-- This query was intended to test "Query that should result in a repartition
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/test/regress/expected/multi_mx_repartition_udt_w1.out]: Will be handled with the following PR.

File: src/test/regress/expected/multi_mx_repartition_udt_w2.out [hunk 14]
```diff
@@ -7,11 +7,12 @@ SET client_min_messages = LOG;
 SET citus.large_table_shard_count = 1;
 SET citus.task_executor_type = 'task-tracker';
 SET citus.log_multi_join_order = true;
--- Query that should result in a repartition join on int column, and be empty.
+-- This query was intended to test "Query that should result in a repartition
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/test/regress/expected/multi_mx_repartition_udt_w2.out]: Will be handled with the following PR.

File: src/test/regress/expected/multi_subquery.out [hunk 15]
```diff
@@ -60,53 +60,6 @@ FROM
...
- l_orderkey |       unit_price       
-------------+------------------------
-        421 |    102749.853333333333
-        806 | 96257.4480681818181818
-        418 | 57151.4156630824373871
-       1124 | 56102.2804738959822181
-        230 | 53847.0509778948909754
-(5 rows)
-
-SELECT
```
[the remark this answers is not in the document]
Comment (velioglu) [reply src/test/regress/expected/multi_subquery.out]: Similar test is added to the multi_subquery_behavioral_analytics file

**Pull request merged**